Develop

DrakeSsq opened the pull request.

DrakeSsq committed:
calculate hours

DrakeSsq committed:
add violations

DrakeSsq committed:
generating a report in json

DrakeSsq committed:
add junit5

DrakeSsq committed:
adding error handling and minor improvements

rustamscode commented on pom.xml:
```diff
@@ -27,11 +27,29 @@
+    <dependency>
+      <groupId>com.fasterxml.jackson.datatype</groupId>
+      <artifactId>jackson-datatype-jsr310</artifactId>
+      <version>2.13.0</version> <!-- или последняя версия -->
```
Забыл комментарий здесь

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,7 +1,44 @@
+        File file = new File(args[0]);
+
+        if (!file.exists()) {
+            throw new FileNotFoundException("file not found");
```
"file not found" нужно вынести в константу. Никогда не оставляй хардкод, выноси в константы.
Сделай сообщение более информативным. `String.format("File %s not found", fileName)` - подобное

rustamscode replied:
Для констант можно завести отдельные неизменяемый класс. Например, `ExceptionMessageUtil `и поместить его в пакет util. 

Lombok немного упрощает это, там есть аннотация [@UtilityClass](https://projectlombok.org/features/experimental/UtilityClass)

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,7 +1,44 @@
+
+        if (!file.exists()) {
+            throw new FileNotFoundException("file not found");
+        } else {
```
Нету смысла делать else. Если у тебя сверху в IF броситься исключение, то сюда уже никогда просто и не дойдет программа

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,7 +1,44 @@
+
+            ObjectMapper objectMapper = new ObjectMapper();
+            objectMapper.registerModule(new JavaTimeModule());
+            objectMapper.enable(SerializationFeature.INDENT_OUTPUT);
```
Можно вынести в отдельный метод статический создание и настройку `ObjectMapper`, чтобы общий поток был чище

rustamscode commented on src/main/java/com/app/exceptions/CustomJsonException.java:
JsonProcessingException достаточно, для чего кастомное исключение?

rustamscode commented on src/main/java/com/app/services/ReportService.java:
```diff
@@ -0,0 +1,69 @@
+        AviationReport aviationReport = new AviationReport();
+        List<SpecialistReport> specialistReports = new ArrayList<>();
+
+
```
2 переноса строки много. Делай 1

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,7 +1,44 @@
+                objectMapper.writeValue(new File("output.json"), aviationReport);
+
+            } catch (JsonProcessingException e) {
+                throw new CustomJsonException("The incoming file contains an unsupported data format");
```
Нужно добавить логирование в `catch` блоках

rustamscode replied:
Message в константы

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,7 +1,44 @@
+        File file = new File(args[0]);
+
+        if (!file.exists()) {
+            throw new FileNotFoundException("file not found");
```
Добавить логирование в catch

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,7 +1,44 @@
+                ReportService reportService = new ReportService(timeService);
+                AviationReport aviationReport = reportService.generateReport();
+
+                objectMapper.writeValue(new File("output.json"), aviationReport);
```
"output.json" нужно в константы

rustamscode commented on src/main/java/com/app/models/Flight.java:
```diff
@@ -2,18 +2,12 @@
-@Setter
-@NoArgsConstructor
-@AllArgsConstructor
+@Data
```
`@Data` никогда не ставь над сущностями. 
Прочитай эту статью https://habr.com/ru/companies/haulmont/articles/836018/

rustamscode commented on src/main/java/com/app/models/Flight.java:
```diff
@@ -2,18 +2,12 @@
+@Data
 public class Flight {
 
     @JsonProperty("aircraft_type")
```
Можно aircraft_type в enum, но так тоже норм

rustamscode commented on src/main/java/com/app/models/Specialist.java:
```diff
@@ -1,20 +1,32 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@Data
```
`@Data` нельзя ставить над сущностями

rustamscode commented on src/main/java/com/app/models/Specialist.java:
```diff
@@ -1,20 +1,32 @@
+            monthlyHours.put(month, monthlyHours.get(month) + hours);
+        } else {
+            monthlyHours.put(month, hours);
+        }
```
Можно использовать новые методы мапы
`monthlyHours.merge(month, hours, Double::sum);`

> Метод merge() может быть объяснен следующим образом: он либо помещает новое значение под заданным ключом (если отсутствует), либо обновляет существующий ключ с заданным значением (UPSERT).

https://mygpstools.com/mapmerge-metod-chtoby-upravlyat-vsemi-ostalnymi

rustamscode commented on src/main/java/com/app/services/ReportService.java:
```diff
@@ -0,0 +1,69 @@
+            for (Map.Entry<Month, Double> entry : spec.getMonthlyHours().entrySet()) {
+                MonthlyReport monthlyReport = new MonthlyReport();
+                monthlyReport.setPeriod(String.valueOf(YearMonth.of(
+                        timeService.getAviationData().getFlights().get(0).getLandingTimeUtc().getYear(),
```
Если getFlights() вернет пустой список, вызов .get(0) приведет к IndexOutOfBoundsException
Нужно делать проверку на null

rustamscode commented on src/main/java/com/app/services/ReportService.java:
```diff
@@ -0,0 +1,69 @@
+
+    }
+
+    public AviationReport generateReport() {
```
Метод `generateReport `слишком много в себе содержит, нарушает принцип единственной ответственности:
получает данные, преобразует специалистов, преобразует месячные отчеты для каждого специалиста, собирает нарушения. Вынеси логику некоторых операций в отдельные методы

rustamscode replied:
Почитать про SOLID, понять суть, уметь привести примеры
https://habr.com/ru/articles/688530/
https://www.baeldung.com/solid-principles

rustamscode commented on src/main/java/com/app/services/ReportService.java:
```diff
@@ -0,0 +1,69 @@
+
+
+                List<String> violations = new ArrayList<>();
+                if (spec.getMonthlyHoursViolation().getOrDefault(entry.getKey(), false)) {
```
Молодец, что использовал `getOrDefault`, аккуратно получилось

rustamscode commented on src/main/java/com/app/services/TimeService.java:
```diff
@@ -0,0 +1,121 @@
+import java.util.Map;
+
+@Getter
+public class TimeService {
```
Связывать service с объектом, состояние которого от него зависит - плохая практика. Сервис должен быть stateless, он только выполняет что-то и отдает результат

rustamscode replied:
Почитать: https://martinfowler.com/eaaCatalog/serviceLayer.html

rustamscode reviewed (commented):
Общие советы:
- Часто используешь циклы, старайся, по возможности заменять их стримами там, где это логично
- Обращай внимание на сложность алгоритмов, которые ты пишешь. Например, очень много кода, где один цикл в другом или даже 3 цикла вложенных. Это очень медленно и будет сыпаться на больших данных.

DrakeSsq committed:
[CR-1] Revision after code-review

-Adding logging
-Following the principles of SOLID
-Correct exception handling
-Most of the code is written in the Stream Api and not in loops

rustamscode commented on src/main/java/com/app/Main.java:
```diff
@@ -1,44 +1,65 @@
+
+        if (args.length == 0) {
+            logger.error(LOG_NO_ARGS);
+            throw ExceptionMessages.createIllegalArgumentException();
```
`throw ExceptionMessages.createIllegalArgumentException();`
Интересный подход, молодец

rustamscode commented on src/main/java/com/app/services/ReportService.java:
```diff
@@ -10,60 +11,72 @@
+
+    private String getReferenceYear(AviationData aviationData, Map.Entry<Month, Double> entry) {
 
+        return Optional.ofNullable(aviationData.getFlights())
```
Супер!
Хорошая идея использовать здесь Optional, тк кроме проверки на null, там еще много другой логики

rustamscode commented on src/main/java/com/app/services/TimeService.java:
```diff
@@ -14,82 +13,67 @@
-    public TimeService(AviationData aviationData) {
-        this.aviationData = aviationData;
-    }
+    public void calculateTime(AviationData aviationData) {
```
Молодец, что развязал AviationData от сервиса. Как вариант можно потом aviationData этот возвращать. Так тоже можно оставить

rustamscode commented on src/main/java/com/app/util/ExceptionMessages.java:
В целом норм. Однако в будущем разделяй константы и exceptions.
Вообще в разных пакетах храни их.
`../util` - здесь создавай какие-то вспомогательные классы или классы для констант
`../exception` - здесь создавай кастомные ошибки

rustamscode reviewed (commented):
Понравилось, что использовал функциональщину в нужных местах. 
Доп чтение для тебя
https://www.baeldung.com/java-factory-pattern
https://habr.com/ru/articles/465835/
